Klustered teams 3 - Skyscanner notes (#10)
rawkode/klustered: notes/klustered-teams/episode-3/skyscanner/README.md

# Kluster 000

## Discovered Symptoms

## Contributing Factors

## Notes from Kluster Breaker

### APIServer Break

> We modified the core `v1.` [apiservice](https://kubernetes.io/docs/tasks/extend-kubernetes/configure-aggregation-layer/#register-apiservice-objects) object to point to the `kubernetes` service in the `default` namespace, resulting in the APIServer being unable to ever successfully break out of a loop and respond as all API requests got trapped in a loop. The reason restarting the APIServer was enough to fix this was that the built-in aggregator server which manages aggregation of APIs within kubernetes rebuilds this list on startup from the API resources defined within the Kube-APIServer.

Reviewer: More than breaking out of the loop, wasn't because the API server can't serve those requests over the service? There is probably a difference between `Local` and the service we added.

Then I have a small suggestion
```quote
Restarting the APIServer fixed the issue because the built-in aggregator server - which manages aggregation of APIs within kubernetes - rebuilds this list on startup from the API resources defined within the Kube-APIServer.
```

Author: You're right, yep.

I've just run through this with a kind cluster again with some verbose logging and the endpoints for the kubernetes service in the default namespace resolve to the same IP and port that Kubernetes is advertising, but the APIserver first gets stuck trying to do OpenAPI discovery against the new endpoints and proceeds to have many TLS Errors:

`error trying to reach service: x509: certificate signed by unknown authority` after a bit of this, it forces a resync of all the endpoints, still fails, and is then unable to renew its lease:

`unable to create required kubernetes system namespace kube-node-lease: the server is currently unable to handle the request (post namespaces)`

This is the point where it starts returning the `service unavailable` we saw.

Author: I've updated this section now, what do you think?

### Cilium/Networking Break

> Cilium is able to replace [kube-proxy](https://kubernetes.io/docs/concepts/overview/components/#kube-proxy) when configured correctly, taking over the maintenance of IPTable rules to allow services within the cluster to be successfully reached. Here we both updated the Cilium config to disable this behaviour, without installing kube-proxy to take over, but also removed the lifecycle related healthchecks from the Cilium Daemonset to enable the new pods to launch successfully and appear to be healthy.

Reviewer: ```suggestion
Cilium is able to replace [kube-proxy](https://kubernetes.io/docs/concepts/overview/components/#kube-proxy) when configured correctly, taking over the maintenance of `iptables` rules to allow services within the cluster to be successfully reached. Here we made two changes:

- we updated the Cilium config to use `kube-proxy` without installing it
- we removed the lifecycle related healthchecks from the Cilium Daemonset to enable the new pods to launch successfully and appear to be healthy.
```

One of the difficult things here is just how many options there are for configuring Cilium due to its power.

### ValidatingAdmissionWebhook

> Nothing clever here, we just created a new [ValidatingAdmissionWebhook](https://kubernetes.io/docs/reference/access-authn-authz/admission-controllers/) for the creation or update of pods which pointed at a random URL and had its `failurePolicy` set to fail. This would have blocked the creation of new pods as the klustered deployment was updated if it hadn't already been deleted.

Reviewer: To be fair, we didn't specify a `failurePolicy` because by default is set to fail.

Reviewer: If we want the yaml

```yaml
apiVersion: admissionregistration.k8s.io/v1
kind: ValidatingWebhookConfiguration
metadata:
  name: default.k8s.io
webhooks:
- name: default.k8s.io
  rules:
  - apiGroups:   [""]
    apiVersions: ["v1"]
    operations:  ["CREATE", "UPDATE"]
    resources:   ["pods"]
    scope:       "Namespaced"
  clientConfig:
    url: https://www.google.com
  admissionReviewVersions: ["v1", "v1beta1"]
  sideEffects: None
  timeoutSeconds: 5
 ```

Reviewer: Nice to point out that `https://www.google.com` was returning error 413 that might have been a little too misleading.

Author: Added, and updated, as appropriate.

# Kluster 000

## Discovered Symptoms

## Contributing Factors

## Notes from Kluster Breaker

### APIServer Break

We modified the core `v1.` [apiservice](https://kubernetes.io/docs/tasks/extend-kubernetes/configure-aggregation-layer/#register-apiservice-objects) object to point to the `kubernetes` service in the `default` namespace, resulting in the APIServer being unable to ever successfully break out of a loop and respond as all API requests got trapped in a loop. The reason restarting the APIServer was enough to fix this was that the built-in aggregator server which manages aggregation of APIs within kubernetes rebuilds this list on startup from the API resources defined within the Kube-APIServer.

> 

Reviewer 2: Might be worth mentioning that this did work on older versions of kubernetes, but didn't here.

```suggestion
Before breaking this, we had only changed `v1.apps` to point to a fake `kubrenetes` (with typo) service in the default namespace. In older versions of kubernetes, we found that pointing to the real `kubernetes` service worked. It goes to show that there are ways you can apply objects that break your cluster (sometimes beyond repair).
```

> ### Cilium/Networking Break

Reviewer 2: Is it also worth putting some yaml in there. Might make it easier for people to understand?

```suggestion

Here is the change that we made to the ApiService:

\`\`\`diff
 apiVersion: apiregistration.k8s.io/v1
 kind: APIService
 metadata:
   labels:
     kube-aggregator.kubernetes.io/automanaged: onstart
   name: v1.
 spec:
   groupPriorityMinimum: 18000
+  service:
+    name: kubernetes
+    namespace: default
   version: v1
   versionPriority: 1
\`\`\`

### Cilium/Networking Break
```

Author: Now added.

# Kluster 000

## Discovered Symptoms

## Contributing Factors

> 

Reviewer 2: Are we going to remove these or do a separate PR on how evil DO were to us?
